BrainPortal/app/models/s3_flat_data_provider.rb

```diff
@@ -0,0 +1,299 @@
+#
+# CBRAIN Project
+#
+# Copyright (C) 2008-2012
+# The Royal Institution for the Advancement of Learning
+# McGill University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Implements a DataProvider that stores its files
+# in a Amazon S3 bucket. Single files are saved
+# as-is, FileCollections are tar'ed and untar'ed
+# as needed.
+class S3FlatDataProvider < DataProvider
+
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  validates_presence_of :cloud_storage_client_identifier, :cloud_storage_client_token,
+                        :cloud_storage_client_path_start, :cloud_storage_client_bucket_name
+  validates :cloud_storage_client_identifier,  length: { is: 20 }
+  validates :cloud_storage_client_token,       length: { is: 40 }
+  validates :cloud_storage_client_bucket_name, format: {
+    with: /\A([a-z]|(d(?!d{0,2}.d{1,3}.d{1,3}.d{1,3})))([a-zd]|(.(?!(.|-)))|(-(?!.))){1,61}[a-zd.]\z/,
+    message: "invalid S3 bucket name, for rules see https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html"
+  }
+
+
+  # This returns the category of the data provider
+  def self.pretty_category_name #:nodoc:
+    "Cloud"
+  end
+
+  # Connects to the S3 service using :cloud_storage_client_identifier and :cloud_storage_client_token;
+  # the connection is maintained in a instance variable!
+  def init_connection
+    @s3_connection = S3Sdkv3Connection.new(self.cloud_storage_client_identifier,
+                                           self.cloud_storage_client_token,
+                                           self.cloud_storage_client_bucket_name,
+                                           self.cloud_storage_client_path_start)
+    @s3_connection.create_bucket(@s3_connection.bucket_name) unless @s3_connection.bucket_exists?(@s3_connection.bucket_name)
```

Comment:
These two lines mean whenever we perform ANY action, we test again and again for the bucket, and we also create and recreated the dummy placeholder file? Since these things only need to be performed ONCE in the lifetime of the DP, why not put them in the `is_alive?()` method instead?